File: code/ex79.java
class Solution {
    public int m = 0, n = 0;
    
    public boolean exist(char[][] board, String word) {
        if (word.length() == 0)    return true;
        m = board.length;    if (m == 0)    return false;
        n = board[0].length;
        
        boolean[][] visited = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, visited, i, j))    return true;
            }
        }
        
        return false;
    }
    
    private boolean dfs(char[][] board, String word, boolean[][] visited, int i, int j) {
        if (word.length() == 1)    return word.charAt(0) == board[i][j] ? true : false;
        if (board[i][j] != word.charAt(0))    return false;
        
        char cur = word.charAt(0);
        visited[i][j] = true;
        
        boolean res = false;
        // Go four directions if possible
        // Go right
        if (j + 1 < n && !visited[i][j + 1])    res |= dfs(board, word.substring(1), visited, i, j + 1);
        if (res)    {visited[i][j] = false;    return true;}
        // Go down
        if (i + 1 < m && !visited[i + 1][j])    res |= dfs(board, word.substring(1), visited, i + 1, j);
        if (res)    {visited[i][j] = false;    return true;}
        // Go up
        if (i - 1 >= 0 && !visited[i - 1][j])    res |= dfs(board, word.substring(1), visited, i - 1, j);
        if (res)    {visited[i][j] = false;    return true;}
        // Go left
        if (j - 1 >= 0 && !visited[i][j - 1])    res |= dfs(board, word.substring(1), visited, i, j - 1);
        if (res)    {visited[i][j] = false;    return true;}
        
        
        visited[i][j] = false;
        
        return false;
    }
}

File: code/ex79.cpp
class Solution {
    int m, n;
    int dir[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
public:
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size();    n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                visited[i][j] = 1;
                if (dfs(board, visited, word, i, j, 0))
                    return 1;
                visited[i][j] = 0;
            }
        }
        return 0;
    }
    
private:
    bool dfs(vector<vector<char>>& board, vector<vector<bool>>& visited, string word, int i, int j, int s) {
        if (s == word.length() - 1 && word[s] == board[i][j])    return 1;
        if (word[s] != board[i][j])    return 0;
        for (auto& d : dir) {
            int ni = i + d[0], nj = j + d[1];
            if (ni < 0 || nj < 0 || ni >= m || nj >= n || visited[ni][nj])    continue;
            visited[ni][nj] = 1;
            if (dfs(board, visited, word, ni, nj, s + 1))    return 1;
            visited[ni][nj] = 0;
        }
        return 0;
    }
};